Tech stacks (#1016)
Codecademy/docs: content/general/concepts/tech-stacks/tech-stacks.md

---
Title: 'Tech Stacks'
Description: 'Tech stacks refer to a set of softwares that work together digital products or solutions such as websites or mobile applications.'
Subjects:
  - 'Computer Science'
  - 'Code Foundations'
Tags:
  - 'ASP.NET'
  - 'Frameworks'
  - 'Rails'
CatalogContent:
  - 'learn-rails'
  - 'paths/full-stack-engineer-career-path'
---

> A **technology (or "tech") stack** refers to a set of softwares that work together to build digital products or solutions such as websites and mobile applications. Essential for building maintainable and scalable software, tech stacks are also referred to as technology infrastructures or solution stacks.

Reviewer: ```suggestion

A **technology (or "tech") stack** refers to a set of softwares that work together to build digital products or solutions such as websites and mobile applications. Essential for building maintainable and scalable software, tech stacks are also referred to as technology infrastructures or solution stacks.
```

### Front-End Tech Stack

The [front-end](https://www.codecademy.com/resources/docs/general/front-end) is the part of a software program that a user interacts with. The [UI/UX design](https://www.codecademy.com/resources/docs/uiux) of an application or website should be seamless and straightforward, which is why it is a primary goal of front-end development.

The three major elements that make up the front-end stack are [HTML](https://www.codecademy.com/resources/docs/html), [CSS](https://www.codecademy.com/resources/docs/css), and [JavaScript](https://www.codecademy.com/resources/docs/javascript).

> ### Back-End Tech Stack

Reviewer: ```suggestion
### Back-End Tech Stack

```

## Common Stacks

### LAMP Stack

The LAMP stack is a popular open-source tech stack that is used to create web apps. It is one of the oldest tech stacks and provides a collection of software for delivering high-performance web applications. This stack comprises the following technologies:

- Linux: An open-source operating system based on the Linux kernel
- Apache: A free and open-source cross-platform web server.
- MySQL: A relational database management system for storing and manipulating data.
- PHP: A server-side scripting programming language used for creating dynamic web applications.

### MERN And Alternative Stacks

MERN is used to describe a specific set of JavaScript-based technologies that are used in the web application development process. It is made of technologies such as MongoDB, Express, React, and Node. The MERN stack is one of the most popular technology stacks. Alternative stacks are the MEVN and MEAN stack. The only difference is Vue.js or Angular.js is used in place of React.js as the front-end framework.

> - MongoDB: MongoDB is a document database used to build highly available and scalable internet applications.

Reviewer: ```suggestion
- MongoDB: A document database used to build highly available and scalable internet applications.
```

The three major elements that make up the front-end stack are [HTML](https://www.codecademy.com/resources/docs/html), [CSS](https://www.codecademy.com/resources/docs/css), and [JavaScript](https://www.codecademy.com/resources/docs/javascript).

### Back-End Tech Stack

The[ back-end](https://www.codecademy.com/resources/docs/general/back-end), or server-side, refers to parts of an application's code that allow it to operate and handle new and existing data. This cannot usually be accessed by a user. The back-end of a computer system is where the majority of data and operating syntax are stored and accessed.

The back-end tech stack is usually comprised of programming languages, [frameworks](https://www.codecademy.com/resources/docs/general/framework), [databases](https://www.codecademy.com/resources/docs/general/database), and [servers](https://www.codecademy.com/resources/docs/general/server).

## Common Stacks

### LAMP Stack

> The LAMP stack is a popular back-end tech stack that is used to create web apps. It is one of the oldest tech stacks and provides a collection of software for delivering high-performance web applications. This stack comprises the following technologies:

Reviewer 2: "The LAMP stack is a popular back-end tech stack ~that is~ used to create web apps. This stack ~comprises~ contains the following technologies:"

- Linux: An open-source operating system based on the Linux kernel
- Apache: A free and open-source cross-platform web server.

> - MySQL: A relational database management system for storing and manipulating data on the backend.

Reviewer 2: "MySQL: A relational database management system for storing and manipulating data on the ~backend~ back-end."

> - PHP: A server-side scripting programming language used for creating dynamic web applications.

Author: The lamp stack is a backend stack that's why I couldn't mention which is frontend

> - PHP: A server-side scripting programming language used for creating dynamic web applications.

Reviewer 3: Then maybe mention this in the introductory paragraph.

### MERN And Alternative Stacks

> MERN is used to describe a specific set of JavaScript-based technologies that are used in the web application development process. It is made of technologies such as MongoDB, Express, React, and Node. The MERN stack is one of the most popular technology stacks. Alternative stacks are the MEVN and MEAN stack. The only difference is Vue.js or Angular.js is used in place of React.js as the front-end framework.

Reviewer 2: ~MERN is used to describe a specific set of JavaScript-based technologies that are used in the web application development process. It is made of technologies such as MongoDB, Express, React, and Node.~

"MERN is a set of JavaScript-based technologies used in the web application development process. MERN is the collection of the following technologies:
- MongoDB: A document database used to build highly available and scalable internet applications.
- Express (.js): A ~backend~ back-end framework built on Node.js ~and it~ providing features for developing mobile and web applications.
- [React(.js)](https://www.codecademy.com/resources/docs/react) / Vue.js / Angular.js : These are ~frontend~ front-end javascript frameworks for building user interfaces and single-page applications.
- Node(.js) is a runtime environment that runs JavaScript on the server-side." 

"The only difference ~is~ being Vue.js or Angular.js is used in place of React.js as the front-end framework.

> 

Reviewer 2: The following definitions should be made above as shown in a previous comment.

- MongoDB: A document database used to build highly available and scalable internet applications.
- Express(.js): A backend framework built on Node.js and it provides features for developing mobile and web applications.
- [React(.js)](https://www.codecademy.com/resources/docs/react) / Vue.js / Angular.js

> - Node(.js) is a runtime environment that runs JavaScript on the server-side.

Reviewer 3: Again, it would be nice to mention front-end vs. back-end here.

Author: I think I've explained this from my last commit. Express is the backend framework while react/vue/angular are the frontend frameworks.

Reviewer 3: Am I missing where it says react/vue/angular are the frontend frameworks?

A tech stack refers to tools, programming languages, and technologies that work together to build digital products or solutions such as websites, mobile, and web apps.

### Frontend Tech Stack
The front end is the part of a software program that a user interacts with. One of the primary goals of front-end development is to create a smooth or frictionless user experience. Basically, the user interface of an application or website should be simple and straightforward.

There are three major elements that make up the frontend stack. These are

- HTML (Hypertext Markup Language): HTML is the standard markup language for building Web pages. It is also the building block of the web and is used to structure a web page and its content.

- CSS (Cascading style sheets): CSS is the language we use to style an HTML document. It describes how HTML elements should be displayed.

- Javascript: JavaScript is the programming language of the Web. It is used to make web pages interactive and it allows the implementation of dynamic web pages. Some of the popular javascript frameworks include React, Vue, Angular and svelte.

### Backend Tech Stack

> The back end also known as the server-side refers to parts of a computer application or a program's code that allow it to operate and that cannot be accessed by a user. The back end of a computer system is where the majority of data and operating syntax are kept and accessible.
> 
> The backend tech stack comprises the following elements:

Reviewer: ```suggestion
The[ back-end](https://www.codecademy.com/resources/docs/general/back-end), or server-side, refers to parts of an application's code that allow it to operate and handle new and existing data. This cannot usually be accessed by a user. The back-end of a computer system is where the majority of data and operating syntax are stored and accessed.

The back-end tech stack is usually comprised of programming languages, [frameworks](https://www.codecademy.com/resources/docs/general/framework), [databases](https://www.codecademy.com/resources/docs/general/database), and [servers](https://www.codecademy.com/resources/docs/general/server).
```

  Programming Languages: A programming language is a computer language programmers use to develop software programs or other sets of instructions for computers to execute. Examples include Javascript, Python, Java, etc.

- Frameworks: A framework is a tool that provides ready-made components or solutions that are customized in order to speed up development. Popular frameworks used in development include Laravel, Ruby on Rails, Express, and Django.

- Databases: A database is a systematic collection of structured information or data. The most common databases include MongoDB, Postgresql, MySQL, and SQL.

> - Servers: A server is a computer program or device that provides a service to another computer program and its user, also known as the client. It responds and accepts to requests made over a network. Commonly used servers are AWS, Google Cloud, Apache, Nginx, Microsoft's Internet Information Services (IIS), etc.

Reviewer: Any chance we could move/incorporate this information into the existing general [back-end entry](https://www.codecademy.com/resources/docs/general/back-end)?

> 
> 
> ### LAMP Stack

Reviewer: ```suggestion

## Common Stacks

### LAMP
```

> The LAMP stack is a popular open-source tech stack that is used to create web apps. It is one of the oldest tech stacks. The LAMP stack provides a collection of software for delivering high-performance web applications. This stack comprises the following technologies

Reviewer: ```suggestion
The LAMP stack is a popular open-source tech stack that is used to create web apps. It is one of the oldest tech stacks and provides a collection of software for delivering high-performance web applications. This stack comprises the following technologies:
```

> - L — Linux: An open-source operating system based on the Linux kernel
> - A — Apache: A free and open-source cross-platform web server.
> - M — MySQL: A relational database management system for storing and manipulating data.
> - P — PHP: A server-side scripting programming language used for creating dynamic web applications.

Reviewer: ```suggestion
- Linux: An open-source operating system based on the Linux kernel
- Apache: A free and open-source cross-platform web server.
- MySQL: A relational database management system for storing and manipulating data.
- PHP: A server-side scripting programming language used for creating dynamic web applications.
```

> ### MERN Stack

Reviewer: For lines # 39 through 63, I think we're repeating some content when describing MongoDB, Express, and Node.js. Could we consolidate the three "ME_N" sections into one section entitled "MERN And Alternative Stacks" that has just one unordered list describing the different softwares used in the stack? Then when talking about MEAN and MEVN we could simply mention that Angular and Vue are the front-end technologies used, respectively.

MERN is used to describe a specific set of JavaScript-based technologies that are used in the web application development process. It is made of technologies such as MongoDB, Express, React, and Node. The MERN stack is one of the most popular technology stacks.

> - M – MongoDB: MongoDB is a document database used to build highly available and scalable internet applications.
> - E – Express(.js): A backend framework built on Nodejs and it provides features for developing mobile and web applications.
> - R — React(.js):  A frontend javascript library developed by Meta for building user interface.
> - N — Node(.js): A javaScript runtime environment that runs on a JavaScript Engine.

Reviewer: ```suggestion
- MongoDB is a document database used to build highly available and scalable internet applications.
- Express(.js) is a back-end framework built on Node.js and it provides features for developing mobile and web applications.
- [React(.js)](https://www.codecademy.com/resources/docs/react) is a front-end library developed by Meta for building user interface.
- Node(.js) is a runtime environment that runs JavaScript on the server-side.
```

We do have existing entries for Express.js and Node.js but I'm not going to suggest linking them at this time because they are in the process of the being moved and having their URLs changed.

- E – Express(.js): A backend framework built on Nodejs and it provides features for developing mobile and web applications.
- V — Vue(.js):  A frontend Javascript framework for building user interfaces and single-page applications.
- N — Node(.js):  A javaScript runtime environment that runs on a JavaScript Engine.
  
### MEAN Stack

The Mean stack is a collection of javascript-based technologies in the development of web apps. The mean stack is used to create full-stack applications. This means that javascript is used across the entire system, from the frontend to the backend to the server and finally to the database. MEAN stands for MongoDB, Express, Angular, and Node.

- M – MongoDB: MongoDB is a document database used to build highly available and scalable internet applications.
- E – Express(.js):  A backend framework built on Nodejs and it provides features for developing mobile and web applications.
- A — Angular(.js): A frontend javascript framework created by Google for building single-page client applications 
- N — Node(.js):  A JavaScript runtime environment that runs on a JavaScript Engine.

> 

Author: Do I take out this part from my entry?
